### PR TITLE
Outline for Active State Transition Function

### PR DESCRIPTION
**Pull request by terenc3t**
Second part of implementing the active transition function. In this PR, I implemented the skeleton code for `computeActiveState`, listed a bunch of TODOs, and plug `computeActiveState` back to beacon chain service. The goal of this PR is to synchronize with block processing logic which I am working together with @rauljordan. Will open issue for each TODO in `computeActiveState`.

### CODE REVIEW

File: beacon-chain/blockchain/core.go [hunk 1]
```diff
@@ -90,6 +91,39 @@ func (b *BeaconChain) persist() error {
 	return b.db.Put([]byte(stateLookupKey), encodedState)
 }
 
+// computeNewActiveState computes a new active state for every beacon block.
+func (b *BeaconChain) computeNewActiveState(seed common.Hash) (*types.ActiveState, error) {
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: So this uses the latest beacon block ?

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -37,3 +45,28 @@ func (c *ChainService) Stop() error {
 	log.Info("Stopping blockchain service")
 	return nil
 }
+
+// updateActiveState receives a beacon block, computes a new active state and writes it to db
+func (c *ChainService) updateActiveState() {
+	for {
+		select {
+		case block := <-c.beaconBlock:
+			log.Debugf("Received beacon block, last active state hash: %v", block.ActiveStateHash)
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: Use the new structure of `log.WithFields(logrus.Fields{"activeStateHash": block.ActiveStateHash}).Debug("Received beacon block")`

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -37,3 +45,28 @@ func (c *ChainService) Stop() error {
 	log.Info("Stopping blockchain service")
 	return nil
 }
+
+// updateActiveState receives a beacon block, computes a new active state and writes it to db
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: Punctuation at end of comments

File: beacon-chain/blockchain/service.go [hunk 4]
```diff
@@ -2,25 +2,32 @@ package blockchain
...
-	ctx      context.Context
-	cancel   context.CancelFunc
-	beaconDB *database.BeaconDB
-	chain    *BeaconChain
+	ctx         context.Context
+	cancel      context.CancelFunc
+	beaconDB    *database.BeaconDB
+	chain       *BeaconChain
+	web3Service *powchain.Web3Service
+	beaconBlock chan *types.Block
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: What is this? Is this when a newly proposed block is received? Can we rename to latestBeaconBlock?

File: beacon-chain/blockchain/core_test.go [hunk 5]
```diff
@@ -143,4 +143,4 @@ func TestShuffle(t *testing.T) {
 	if reflect.DeepEqual(list1, list2) {
 		t.Errorf("2 shuffled lists shouldn't be equal")
 	}
-}
+}
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/core_test.go]: Please revert these removals of line ends. It creates unnecessary diffs

File: beacon-chain/blockchain/core.go (hunk 1 above)
Comment (terenc3t) [reply beacon-chain/blockchain/core.go]: Correct!

File: beacon-chain/utils/shuffle_test.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+	list1, err := Shuffle(hash1, 100)
+	if err != nil {
+		t.Errorf("Shuffle failed with: %v", err)
+	}
+
+	list2, err := Shuffle(hash2, 100)
+	if err != nil {
+		t.Errorf("Shuffle failed with: %v", err)
+	}
+	if reflect.DeepEqual(list1, list2) {
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle_test.go]: What is this actually testing?

Does shuffle guarantee that the output will never be the same order as the input? 
Random shuffle could return the same output as the input if it is truly random. 

Also, list1 and list2 consist of completely different inputs so they would never equal the same even if Shuffle did nothing.

Comment (terenc3t) [reply beacon-chain/utils/shuffle_test.go]: It basically test the following algorithm in the spec which implemented by other sharding team: 
https://notes.ethereum.org/SCIg8AH5SA-O4C1G1LYZHQ?view#Helper-functions

it's testing the 2 lists using different seed are going to have different index orders. I haven't figured out a better way to test this algorithm. Let me know if you have something in mind...

Comment (prestonvanloon) [reply beacon-chain/utils/shuffle_test.go]: Oh, I misunderstood what shuffle does. I thought it shuffled the elements in the list, but it actually returns the indices in shuffled order.

What do you think about calling this `ShuffleIndices` for clarity?

File: beacon-chain/utils/shuffle.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	hashSeed, err := blake2s.New256(seed[:])
+	if err != nil {
+		return nil, err
+	}
+
+	hashSeedByte := hashSeed.Sum(nil)
+
+	// shuffle stops at the second to last index
```
Comment (rauljordan) [inline beacon-chain/utils/shuffle.go]: Punctuation

**Pull request merged**